File: app/domain/portfolio_hash.py
"""
Portfolio Hash Generation

Generates a deterministic hash from current portfolio state.
Used to identify when recommendations apply to the same portfolio state.

The hash includes:
- All positions (including zero quantities for securities in universe)
- Cash balances as pseudo-positions (CASH.EUR, CASH.USD, etc.)
- The full securities universe to detect when new securities are added
- Per-symbol configuration: allow_buy, allow_sell, min_portfolio_target, max_portfolio_target, country, industry
"""

import hashlib
import logging
from typing import Any, Dict, List, Optional

from app.domain.models import Security

logger = logging.getLogger(__name__)


def apply_pending_orders_to_portfolio(
    positions: List[Dict[str, Any]],
    cash_balances: Dict[str, float],
    pending_orders: List[Dict[str, Any]],
) -> tuple[List[Dict[str, Any]], Dict[str, float]]:
    """
    Apply pending orders to positions and cash balances to get hypothetical future state.

    For pending BUY orders: reduces cash balance by quantity * price in the order's currency.
    For pending SELL orders: reduces position quantity by the order quantity.

    Args:
        positions: List of position dicts with 'symbol' and 'quantity' keys
        cash_balances: Dict of currency -> amount (e.g., {"EUR": 1500.0, "USD": 200.0})
        pending_orders: List of pending order dicts with keys: symbol, side, quantity, price, currency

    Returns:
        Tuple of (adjusted_positions, adjusted_cash_balances)

    Example:
        positions = [{"symbol": "AAPL", "quantity": 10}]
        cash = {"EUR": 1500.0}
        orders = [{"symbol": "AAPL", "side": "buy", "quantity": 5, "price": 100.0, "currency": "EUR"}]
        adjusted_pos, adjusted_cash = apply_pending_orders_to_portfolio(positions, cash, orders)
        # adjusted_pos = [{"symbol": "AAPL", "quantity": 15}]
        # adjusted_cash = {"EUR": 1000.0}
    """
    # Create a copy of positions as a dict for easier manipulation
    position_map: Dict[str, float] = {
        p["symbol"].upper(): float(p.get("quantity", 0) or 0) for p in positions
    }

    # Create a copy of cash balances
    adjusted_cash = cash_balances.copy()

    # Process each pending order
    for order in pending_orders:
        symbol = (order.get("symbol") or "").upper()
        side = (order.get("side") or "").lower()
        quantity = float(order.get("quantity", 0) or 0)
        price = float(order.get("price", 0) or 0)
        currency = order.get("currency") or "EUR"

        if not symbol or quantity <= 0 or price <= 0:
            logger.warning(
                f"Skipping invalid pending order: symbol={symbol}, quantity={quantity}, price={price}"
            )
            continue

        if side == "buy":
            # Reduce cash by the order value
            order_value = quantity * price
            current_cash = adjusted_cash.get(currency, 0.0)
            adjusted_cash[currency] = max(0.0, current_cash - order_value)

            # Increase position quantity (assuming order will execute)
            current_quantity = position_map.get(symbol, 0.0)
            position_map[symbol] = current_quantity + quantity

            logger.debug(
                f"Applied pending BUY: {symbol} +{quantity}, cash {currency} -{order_value:.2f}"
            )

        elif side == "sell":
            # Reduce position quantity
            current_quantity = position_map.get(symbol, 0.0)
            new_quantity = max(0.0, current_quantity - quantity)
            position_map[symbol] = new_quantity

            # Note: Cash is not increased here because SELL orders don't generate cash until executed
            # The planner should account for this when calculating available cash

            logger.debug(
                f"Applied pending SELL: {symbol} -{quantity} (from {current_quantity} to {new_quantity})"
            )

        else:
            logger.warning(f"Unknown order side: {side} for order {symbol}")

    # Convert position_map back to list of dicts
    adjusted_positions = [
        {"symbol": symbol, "quantity": int(qty) if qty > 0 else 0}
        for symbol, qty in position_map.items()
        if qty > 0  # Only include positions with quantity > 0
    ]

    return adjusted_positions, adjusted_cash


def generate_portfolio_hash(
    positions: List[Dict[str, Any]],
    securities: Optional[List[Security]] = None,
    cash_balances: Optional[Dict[str, float]] = None,
    pending_orders: Optional[List[Dict[str, Any]]] = None,
) -> str:
    """
    Generate a deterministic hash from current portfolio state.

    Args:
        positions: List of position dicts with 'symbol' and 'quantity' keys
        securities: Optional list of Security objects in universe (to detect new securities and include config)
        cash_balances: Optional dict of currency -> amount (e.g., {"EUR": 1500.0})
        pending_orders: Optional list of pending order dicts with keys: symbol, side, quantity, price, currency

    Returns:
        8-character hex hash (first 8 chars of MD5)

    Example:
        positions = [{"symbol": "AAPL", "quantity": 10}]
        securities = [Security(symbol="AAPL", ...), Security(symbol="MSFT", ...)]
        cash = {"EUR": 1500.0, "USD": 200.0}
        orders = [{"symbol": "AAPL", "side": "buy", "quantity": 5, "price": 100.0, "currency": "EUR"}]
        hash = generate_portfolio_hash(positions, securities, cash, orders)
    """
    # Apply pending orders to get hypothetical future state
    if pending_orders:
        positions, cash_balances = apply_pending_orders_to_portfolio(
            positions, cash_balances or {}, pending_orders
        )

    # Build a dict of symbol -> quantity from positions
    # Use float | int to handle both security quantities (int) and cash (float)
    position_map: Dict[str, float | int] = {
        p["symbol"].upper(): int(p.get("quantity", 0) or 0) for p in positions
    }

    # Build a dict of symbol -> security config data
    stock_config_map: Dict[str, Dict[str, Any]] = {}

    if securities:
        for security in securities:
            symbol_upper = security.symbol.upper()
            # Ensure security is in position_map (with 0 if not held)
            if symbol_upper not in position_map:
                position_map[symbol_upper] = 0

            # Extract config fields
            stock_config_map[symbol_upper] = {
                "allow_buy": security.allow_buy,
                "allow_sell": security.allow_sell,
                "min_portfolio_target": security.min_portfolio_target,
                "max_portfolio_target": security.max_portfolio_target,
                "country": security.country or "",
                "industry": security.industry or "",
            }

    # Add cash balances as pseudo-positions (filter out zero balances)
    if cash_balances:
        for currency, amount in cash_balances.items():
            if amount > 0:
                # Round to 2 decimal places for stability
                position_map[f"CASH.{currency.upper()}"] = round(amount, 2)

    # Sort by symbol for deterministic ordering
    sorted_symbols = sorted(position_map.keys())

    # Build canonical string: "SYMBOL:QUANTITY:allow_buy:allow_sell:min_target:max_target:country:industry"
    parts = []
    for symbol in sorted_symbols:
        quantity = position_map[symbol]
        # Use int for security quantities, float for cash
        if symbol.startswith("CASH."):
            parts.append(f"{symbol}:{quantity}")
        else:
            # Get config for this symbol (use defaults if not in securities list)
            config = stock_config_map.get(symbol, {})
            allow_buy = config.get("allow_buy", True)
            allow_sell = config.get("allow_sell", False)
            min_target = config.get("min_portfolio_target")
            max_target = config.get("max_portfolio_target")
            country = config.get("country", "")
            industry = config.get("industry", "")

            # Format config fields
            min_target_str = "" if min_target is None else str(min_target)
            max_target_str = "" if max_target is None else str(max_target)

            parts.append(
                f"{symbol}:{int(quantity)}:{allow_buy}:{allow_sell}:{min_target_str}:{max_target_str}:{country}:{industry}"
            )

    canonical = ",".join(parts)

    # Generate hash and return first 8 characters
    full_hash = hashlib.md5(canonical.encode()).hexdigest()
    return full_hash[:8]


def generate_settings_hash(settings_dict: Dict[str, Any]) -> str:
    """
    Generate a deterministic hash from settings that affect recommendations.

    Args:
        settings_dict: Dictionary of settings values

    Returns:
        8-character hex hash (first 8 chars of MD5)

    Example:
        settings = {"min_trade_size": 100, "min_security_score": 0.5}
        hash = generate_settings_hash(settings)  # e.g., "b1c2d3e4"
    """
    # Settings that affect recommendation calculations
    # Note: min_trade_size and recommendation_depth removed (handled by optimizer now)
    relevant_keys = sorted(
        [
            "min_security_score",
            "min_hold_days",
            "sell_cooldown_days",
            "max_loss_threshold",
            "target_annual_return",
            "optimizer_blend",
            "optimizer_target_return",
            "transaction_cost_fixed",
            "transaction_cost_percent",
            "min_cash_reserve",
            "max_plan_depth",
        ]
    )

    # Build canonical string: "key:value,key:value,..."
    parts = [f"{k}:{settings_dict.get(k, '')}" for k in relevant_keys]
    canonical = ",".join(parts)

    # Generate hash and return first 8 characters
    full_hash = hashlib.md5(canonical.encode()).hexdigest()
    return full_hash[:8]


def generate_allocations_hash(allocations_dict: Dict[str, float]) -> str:
    """
    Generate a deterministic hash from allocation targets.

    Args:
        allocations_dict: Dictionary of allocation targets with keys like
                         "country:United States" or "industry:Technology"
                         and values as target percentages

    Returns:
        8-character hex hash (first 8 chars of MD5)

    Example:
        allocations = {"country:United States": 0.6, "industry:Technology": 0.3}
        hash = generate_allocations_hash(allocations)  # e.g., "a1b2c3d4"
    """
    if not allocations_dict:
        return "00000000"  # Empty allocations

    # Sort by key for deterministic ordering
    sorted_keys = sorted(allocations_dict.keys())

    # Build canonical string: "key:value,key:value,..."
    # Round values to 4 decimal places for stability
    parts = [f"{k}:{round(allocations_dict[k], 4)}" for k in sorted_keys]
    canonical = ",".join(parts)

    # Generate hash and return first 8 characters
    full_hash = hashlib.md5(canonical.encode()).hexdigest()
    return full_hash[:8]


def generate_recommendation_cache_key(
    positions: List[Dict[str, Any]],
    settings_dict: Dict[str, Any],
    securities: Optional[List[Security]] = None,
    cash_balances: Optional[Dict[str, float]] = None,
    allocations_dict: Optional[Dict[str, float]] = None,
    pending_orders: Optional[List[Dict[str, Any]]] = None,
) -> str:
    """
    Generate a cache key from portfolio state, settings, and allocations.

    This ensures that cache is invalidated when positions, settings,
    securities universe, cash balances, per-symbol configuration, allocation targets,
    or pending orders change.

    Args:
        positions: List of position dicts with 'symbol' and 'quantity' keys
        settings_dict: Dictionary of settings values
        securities: Optional list of Security objects in universe
        cash_balances: Optional dict of currency -> amount
        allocations_dict: Optional dict of allocation targets (e.g., {"country:US": 0.6})
        pending_orders: Optional list of pending order dicts with keys: symbol, side, quantity, price, currency

    Returns:
        26-character combined hash (portfolio_hash:settings_hash:allocations_hash)

    Example:
        positions = [{"symbol": "AAPL", "quantity": 10}]
        settings = {"min_security_score": 0.5}
        securities = [Security(symbol="AAPL", ...), Security(symbol="MSFT", ...)]
        cash = {"EUR": 1500.0}
        allocations = {"country:United States": 0.6}
        orders = [{"symbol": "AAPL", "side": "buy", "quantity": 5, "price": 100.0, "currency": "EUR"}]
        key = generate_recommendation_cache_key(positions, settings, securities, cash, allocations, orders)
    """
    portfolio_hash = generate_portfolio_hash(
        positions, securities, cash_balances, pending_orders
    )
    settings_hash = generate_settings_hash(settings_dict)
    allocations_hash = generate_allocations_hash(allocations_dict or {})
    return f"{portfolio_hash}:{settings_hash}:{allocations_hash}"

File: trader/internal/modules/planning/hash/portfolio_hash.go
package hash

import (
	"crypto/md5"
	"fmt"
	"math"
	"sort"
	"strings"

	"github.com/aristath/arduino-trader/internal/modules/universe"
	"github.com/rs/zerolog/log"
)

// Position represents a portfolio position for hashing
type Position struct {
	Symbol   string
	Quantity int
}

// PendingOrder represents a pending order for hashing
type PendingOrder struct {
	Symbol   string
	Side     string // "buy" or "sell"
	Quantity int
	Price    float64
	Currency string
}

// ApplyPendingOrdersToPortfolio applies pending orders to positions and cash balances
// to get hypothetical future state.
//
// For pending BUY orders: reduces cash balance by quantity * price in the order's currency.
// For pending SELL orders: reduces position quantity by the order quantity.
//
// Args:
//   - positions: List of positions with symbol and quantity
//   - cashBalances: Map of currency -> amount (e.g., {"EUR": 1500.0, "USD": 200.0})
//   - pendingOrders: List of pending orders
//   - allowNegativeCash: If true, allows cash to go negative for hypothetical planning.
//     If false (default), clamps cash at 0 for safety.
//
// Returns:
//   - Adjusted positions and cash balances
func ApplyPendingOrdersToPortfolio(
	positions []Position,
	cashBalances map[string]float64,
	pendingOrders []PendingOrder,
	allowNegativeCash bool,
) ([]Position, map[string]float64) {
	// Create a copy of positions as a map for easier manipulation
	positionMap := make(map[string]int)
	for _, p := range positions {
		symbol := strings.ToUpper(p.Symbol)
		if p.Quantity > 0 {
			positionMap[symbol] = p.Quantity
		}
	}

	// Create a copy of cash balances
	adjustedCash := make(map[string]float64)
	for currency, amount := range cashBalances {
		adjustedCash[currency] = amount
	}

	// Process each pending order
	for _, order := range pendingOrders {
		symbol := strings.ToUpper(order.Symbol)
		side := strings.ToLower(order.Side)
		quantity := order.Quantity
		price := order.Price
		currency := order.Currency
		if currency == "" {
			currency = "EUR"
		}

		if symbol == "" || quantity <= 0 || price <= 0 {
			log.Warn().
				Str("symbol", symbol).
				Int("quantity", quantity).
				Float64("price", price).
				Msg("Skipping invalid pending order")
			continue
		}

		if side == "buy" {
			// Reduce cash by the order value
			orderValue := float64(quantity) * price
			currentCash := adjustedCash[currency]
			newCash := currentCash - orderValue

			// Clamp to zero unless negative cash is explicitly allowed
			if allowNegativeCash {
				adjustedCash[currency] = newCash
			} else {
				adjustedCash[currency] = math.Max(0.0, newCash)
			}

			// Increase position quantity (assuming order will execute)
			currentQuantity := positionMap[symbol]
			positionMap[symbol] = currentQuantity + quantity

			log.Debug().
				Str("symbol", symbol).
				Int("quantity", quantity).
				Str("currency", currency).
				Float64("order_value", orderValue).
				Msg("Applied pending BUY")

		} else if side == "sell" {
			// Reduce position quantity
			currentQuantity := positionMap[symbol]
			newQuantity := int(math.Max(0.0, float64(currentQuantity-quantity)))
			if newQuantity > 0 {
				positionMap[symbol] = newQuantity
			} else {
				delete(positionMap, symbol)
			}

			// Note: Cash is not increased here because SELL orders don't generate cash until executed
			// The planner should account for this when calculating available cash

			log.Debug().
				Str("symbol", symbol).
				Int("quantity", quantity).
				Int("from", currentQuantity).
				Int("to", newQuantity).
				Msg("Applied pending SELL")

		} else {
			log.Warn().Str("side", side).Str("symbol", symbol).Msg("Unknown order side")
		}
	}

	// Convert position_map back to list
	adjustedPositions := make([]Position, 0, len(positionMap))
	for symbol, qty := range positionMap {
		if qty > 0 {
			adjustedPositions = append(adjustedPositions, Position{
				Symbol:   symbol,
				Quantity: qty,
			})
		}
	}

	return adjustedPositions, adjustedCash
}

// GeneratePortfolioHash generates a deterministic hash from current portfolio state.
//
// The hash includes:
//   - All positions (including zero quantities for securities in universe)
//   - Cash balances as pseudo-positions (CASH.EUR, CASH.USD, etc.)
//   - The full securities universe to detect when new securities are added
//   - Per-symbol configuration: allow_buy, allow_sell, min_portfolio_target, max_portfolio_target, country, industry
//
// Args:
//   - positions: List of positions with symbol and quantity
//   - securities: Optional list of Security objects in universe (to detect new securities and include config)
//   - cashBalances: Optional map of currency -> amount (e.g., {"EUR": 1500.0})
//   - pendingOrders: Optional list of pending orders
//
// Returns:
//   - 8-character hex hash (first 8 chars of MD5)
func GeneratePortfolioHash(
	positions []Position,
	securities []*universe.Security,
	cashBalances map[string]float64,
	pendingOrders []PendingOrder,
) string {
	// Apply pending orders to get hypothetical future state
	if len(pendingOrders) > 0 {
		// Allow negative cash for hypothetical portfolio state
		positions, cashBalances = ApplyPendingOrdersToPortfolio(
			positions,
			cashBalances,
			pendingOrders,
			true, // allow_negative_cash
		)
	}

	// Build a map of symbol -> quantity from positions
	positionMap := make(map[string]interface{})
	for _, p := range positions {
		symbol := strings.ToUpper(p.Symbol)
		positionMap[symbol] = p.Quantity
	}

	// Build a map of symbol -> security config data
	stockConfigMap := make(map[string]map[string]interface{})

	if securities != nil {
		for _, security := range securities {
			symbolUpper := strings.ToUpper(security.Symbol)
			// Ensure security is in position_map (with 0 if not held)
			if _, exists := positionMap[symbolUpper]; !exists {
				positionMap[symbolUpper] = 0
			}

			// Extract config fields
			country := security.Country
			industry := security.Industry

			minTarget := ""
			if security.MinPortfolioTarget > 0 {
				minTarget = fmt.Sprintf("%v", security.MinPortfolioTarget)
			}
			maxTarget := ""
			if security.MaxPortfolioTarget > 0 {
				maxTarget = fmt.Sprintf("%v", security.MaxPortfolioTarget)
			}

			stockConfigMap[symbolUpper] = map[string]interface{}{
				"allow_buy":            security.AllowBuy,
				"allow_sell":           security.AllowSell,
				"min_portfolio_target": minTarget,
				"max_portfolio_target": maxTarget,
				"country":              country,
				"industry":             industry,
			}
		}
	}

	// Add cash balances as pseudo-positions (filter out zero balances)
	if cashBalances != nil {
		for currency, amount := range cashBalances {
			if amount > 0 {
				// Round to 2 decimal places for stability
				rounded := math.Round(amount*100) / 100
				positionMap[fmt.Sprintf("CASH.%s", strings.ToUpper(currency))] = rounded
			}
		}
	}

	// Sort by symbol for deterministic ordering
	sortedSymbols := make([]string, 0, len(positionMap))
	for symbol := range positionMap {
		sortedSymbols = append(sortedSymbols, symbol)
	}
	sort.Strings(sortedSymbols)

	// Build canonical string: "SYMBOL:QUANTITY:allow_buy:allow_sell:min_target:max_target:country:industry"
	parts := make([]string, 0, len(sortedSymbols))
	for _, symbol := range sortedSymbols {
		quantity := positionMap[symbol]

		// Use different format for cash vs securities
		if strings.HasPrefix(symbol, "CASH.") {
			parts = append(parts, fmt.Sprintf("%s:%v", symbol, quantity))
		} else {
			// Get config for this symbol (use defaults if not in securities list)
			config := stockConfigMap[symbol]
			if config == nil {
				config = map[string]interface{}{
					"allow_buy":            true,
					"allow_sell":           false,
					"min_portfolio_target": "",
					"max_portfolio_target": "",
					"country":              "",
					"industry":             "",
				}
			}

			part := fmt.Sprintf("%s:%d:%v:%v:%s:%s:%s:%s",
				symbol,
				quantity,
				config["allow_buy"],
				config["allow_sell"],
				config["min_portfolio_target"],
				config["max_portfolio_target"],
				config["country"],
				config["industry"],
			)
			parts = append(parts, part)
		}
	}

	canonical := strings.Join(parts, ",")

	// Generate hash and return first 8 characters
	hash := md5.Sum([]byte(canonical))
	fullHash := fmt.Sprintf("%x", hash)
	return fullHash[:8]
}

// GenerateSettingsHash generates a deterministic hash from settings that affect recommendations.
//
// Args:
//   - settings: Map of settings values
//
// Returns:
//   - 8-character hex hash (first 8 chars of MD5)
func GenerateSettingsHash(settings map[string]interface{}) string {
	// Settings that affect recommendation calculations
	relevantKeys := []string{
		"min_security_score",
		"min_hold_days",
		"sell_cooldown_days",
		"max_loss_threshold",
		"target_annual_return",
		"optimizer_blend",
		"optimizer_target_return",
		"transaction_cost_fixed",
		"transaction_cost_percent",
		"min_cash_reserve",
		"max_plan_depth",
	}
	sort.Strings(relevantKeys)

	// Build canonical string: "key:value,key:value,..."
	parts := make([]string, 0, len(relevantKeys))
	for _, k := range relevantKeys {
		value := ""
		if v, exists := settings[k]; exists && v != nil {
			value = fmt.Sprintf("%v", v)
		}
		parts = append(parts, fmt.Sprintf("%s:%s", k, value))
	}
	canonical := strings.Join(parts, ",")

	// Generate hash and return first 8 characters
	hash := md5.Sum([]byte(canonical))
	fullHash := fmt.Sprintf("%x", hash)
	return fullHash[:8]
}

// GenerateAllocationsHash generates a deterministic hash from allocation targets.
//
// Args:
//   - allocations: Map of allocation targets with keys like "country:United States"
//     or "industry:Technology" and values as target percentages
//
// Returns:
//   - 8-character hex hash (first 8 chars of MD5)
func GenerateAllocationsHash(allocations map[string]float64) string {
	if len(allocations) == 0 {
		return "00000000" // Empty allocations
	}

	// Sort by key for deterministic ordering
	sortedKeys := make([]string, 0, len(allocations))
	for k := range allocations {
		sortedKeys = append(sortedKeys, k)
	}
	sort.Strings(sortedKeys)

	// Build canonical string: "key:value,key:value,..."
	// Round values to 4 decimal places for stability
	parts := make([]string, 0, len(sortedKeys))
	for _, k := range sortedKeys {
		rounded := math.Round(allocations[k]*10000) / 10000
		parts = append(parts, fmt.Sprintf("%s:%.4f", k, rounded))
	}
	canonical := strings.Join(parts, ",")

	// Generate hash and return first 8 characters
	hash := md5.Sum([]byte(canonical))
	fullHash := fmt.Sprintf("%x", hash)
	return fullHash[:8]
}

// GenerateRecommendationCacheKey generates a cache key from portfolio state, settings, and allocations.
//
// This ensures that cache is invalidated when positions, settings,
// securities universe, cash balances, per-symbol configuration, allocation targets,
// or pending orders change.
//
// Args:
//   - positions: List of positions with symbol and quantity
//   - settings: Map of settings values
//   - securities: Optional list of Security objects in universe
//   - cashBalances: Optional map of currency -> amount
//   - allocations: Optional map of allocation targets (e.g., {"country:US": 0.6})
//   - pendingOrders: Optional list of pending orders
//
// Returns:
//   - 26-character combined hash (portfolio_hash:settings_hash:allocations_hash)
func GenerateRecommendationCacheKey(
	positions []Position,
	settings map[string]interface{},
	securities []*universe.Security,
	cashBalances map[string]float64,
	allocations map[string]float64,
	pendingOrders []PendingOrder,
) string {
	portfolioHash := GeneratePortfolioHash(positions, securities, cashBalances, pendingOrders)
	settingsHash := GenerateSettingsHash(settings)
	allocationsHash := GenerateAllocationsHash(allocations)
	return fmt.Sprintf("%s:%s:%s", portfolioHash, settingsHash, allocationsHash)
}
